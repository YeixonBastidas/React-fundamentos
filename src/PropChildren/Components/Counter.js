import React, { useState } from 'react';


export const Counter = ({children}) => {
const [ clicks, setClicks] = useState(0);

    const increment = () => setClicks(clicks + 1);
    const decrement = () => setClicks(clicks - 1);

    if(!children) {
        const styles = {
            background : '#CC6040',
            borderRadius: '0.2em',
            padding: '0.3em 1em',
            color: '#FFF'
        }
        return(
            <div style={styles}>
                Wops!! Debes de pasar componentes como {'<Button />'}
            </div>
        )
    }

    const _componetChildren = React.Children.map(children, (child) => {

        const porps = {};
        if(child.type === Title) {
            porps.clicks = clicks
        }

        if(child.type === Button) {
            porps.onIncrement = increment
            porps.onDecrement = decrement
        }

        return React.cloneElement(child, porps)
    })

    return _componetChildren
}

export const Title = ({ clicks, children }) => {
     return typeof children === 'function'
          ? children(clicks)
          :  <span>{ clicks }</span>
}

export const Button = ({ type, onIncrement, onDecrement }) => {
    const action = () => {
        type === 'increment' ? onIncrement() : onDecrement();
    }

    return (
        <button onClick={action}>
            { type === 'increment' ? 'Agregar' : 'Quitar'}
        </button>
    )
}