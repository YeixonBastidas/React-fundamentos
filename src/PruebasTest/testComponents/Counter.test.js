import React from 'react';
import Enzyme, {shallow} from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'
import Counter from './Counter.js'


// configuramos ezyme para cada archivo de test
Enzyme.configure({ adapter: new Adapter() });

let wrapper;

// cada vez q se llame en los test crea una nueva instancia 
beforeEach(() => {
    wrapper = shallow(<Counter />)
})

describe('Probando componente con estados', () => {

    test('Valdiar que coincida con Snapshot', () => {
        expect(toJson(wrapper)).toMatchSnapshot()
    })

    test('Valdiar funcionamiento de botones', () => {
        wrapper.find('button').first().simulate('click') // creamo click en el boton de incrementar 
        
        expect(wrapper.find('h1').text()).toBe('1') //

        wrapper.find('button').last().simulate('click') // creamo click en el boton de incrementar 
        wrapper.find('button').last().simulate('click')

        expect(wrapper.find('h1').text()).toBe('-1') 

    })    
})