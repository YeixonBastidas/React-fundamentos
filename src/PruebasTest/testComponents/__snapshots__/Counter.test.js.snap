// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando componente con estados Valdiar que coincida con Snapshot 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Increment
  </button>
  -
  <button
    onClick={[Function]}
  >
    Decrement
  </button>
  <h1>
    0
  </h1>
</div>
`;
